aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -1144,7 +1154,9 @@ Tensor& argmax_out(const Tensor& self, c10::optional<int64_t> dim, bool keepdim,
   }
   auto itr = make_reduction("argmax", result, in, dim.value_or(0), keepdim,
       self.scalar_type(), at::kLong);
-  argmax_stub(itr.device_type(), itr);
+  if (self.numel() != 0) {
```

Comment:
Why are you testing self.numel() instead of itr.numel()?

Reply:
That's a very good point actually, maybe we (in this PR) and numpy are doing this wrong, and we should be checking the numel of output to determine if op is allowable. 
In numpy (and in this pr), 
- min(axis=0) for tensor of size (5,0) will succeed and produce tensor of size (0,), 
- same op for tensor of size (0,0) will fail, even though it also could produce the tensor of size (0,). 

numpy error message says "zero-size array to reduction operation which has no identity". There's nothing magical about non-zero reduction size in the first case - there are still no real elements in that nominally non-zero size, and no reduction is happening. It's just that we don't have to care about identity element because output is still empty. But with that reasoning second case should be ok too.

Reply:
> same op for tensor of size (0,0) will fail, even though it also could produce the tensor of size (0,).

That would result in a weird  inconsistency, if `(0, 0) -> (0,)`, what would `(0,)` input give? 

Currently what happens is that a shape a reduction over axis 0 of input shape `(N, M)` gets reduced to `(1, M)`, followed by a squeeze over the axis that is being reduced over, so the end result is of shape `(M,)`. If `N` is of size 0, that first step cannot work.

Currently a reduction over a size-0 dimension always errors out, which is consistent behavior. Special-casing "size-0 doesn't error out if another axis is also of size 0" would be odd.

Reply:
Reduction over size-0 dimension works for ops that have identity (sum, prod, norms that have identity), so apparently first step works for them even for N=0. For those operations, even if the output has real elements, and reduction was over 0 elements, we still know what to fill output with (identity). For min/max related operations, we don't know what to do when output has real elements, so that should be the criterion, instead of "error out when reduction dimension has 0 elements". In most cases (single reduction dimension, single zero dimension) those rules are equivalent, for those cases where they differ, I can't see rationale to go with less permissive one (don't reduce over zero dimension).

Reply:
Regarding @ngimel 's comment above, I have the following to add:

* For arithmetic operations that have an identity (sum/prod/etc), reducing over the dimension that is 0 for `numel() == 0` can be possible since the output will have `numel() == 0` too. However:
  - For the case where `keepdim=True` the same sized tensor will be returned but with the dim 0 as `1`. This means that the result will be a non-zero tensor (unless we want to change the behaviour of the methods for `numel() == 0` and otherwise.
  - For the case where `keepdim=False` the dimension that has `0` will be gone and we still need to return a tensor with `numel() != 0`.
  - In both cases some values will need to be filled into the tensor before returning it.
* The same issue will apply for min/max like operations too.

Are such outputs from size-0 tensors useful for pytorch users?

Reply:
For reference, the identity element of `min` is `infty` and of `max` is `-infty` (when there are no `NaN`s in the tensor at least). On the one hand, I would think that it is a bit odd to return a vector filled with infinities when computing a max or a min.
On the other, it is what is mathematically correct, and some people may count on mathematical consistency. As a counter-argument for the previous point, one may also say that "it's already quite odd to do the min or max over a 0-dim".

Reply:
One thing to consider is that `infinity` and `-infinity` can't be returned for tensors with integral dtype.

Reply:
For reference, these ideas of using the operations `min` with `+infty` as the identity elements are used in the area of [Tropical Geometry](https://en.wikipedia.org/wiki/Tropical_geometry). This area has applications on linear programming, so I would not be surprised if someone wanted to write a library in PyTorch to use it for that.
About integral types, the identity element in that case is `std::numeric_limits<T>::max`.

Reply:
Regarding use of `std::numeric_limits<T>::max` for identity,
Personally, I think it will be odd that identity will change with the dtype (int8 -> 127, int16 -> 32,767, etc).

Reply:
> For min/max related operations, we don't know what to do when output has real elements, so that should be the criterion, instead of "error out when reduction dimension has 0 elements".

Let me make it more explicit:

```python
>>> # The 1-D case
>>> np.max(np.ones((5,)), axis=0)
1.0
>>> np.max(np.ones((0,)), axis=0)
...
ValueError: zero-size array to reduction operation maximum which has no identity

# Append a second dimension that's not being reduced over: this dimension
# always keep the same size, it does not participate in any way in the reduction
>>> np.max(np.ones((5,3)), axis=0)
array([1., 1., 1.])
>>> np.max(np.ones((0,3)), axis=0)
...
ValueError: zero-size array to reduction operation maximum which has no identity

# Now use that same 2-D example, but with size of second dimension being 0
>>> np.max(np.ones((5,0)), axis=0)
array([], dtype=float64)
>>> np.max(np.ones((0,0)), axis=0)
...
ValueError: zero-size array to reduction operation maximum which has no identity
```

> In most cases (single reduction dimension, single zero dimension) those rules are equivalent, for those cases where they differ, I can't see rationale to go with less permissive one (don't reduce over zero dimension).

Both designs are entirely self-consistent, so either choice works. I personally find the one considering only the axis being reduced over more logical (it fits my mental model of "reduce over axis" does). But really it's in practice not very relevant, so I'd just follow what NumPy does here.

Reply:
> Personally, I think it will be odd that identity will change with the dtype (int8 -> 127, int16 -> 32,767, etc).

Agreed, this is not a good idea, and it will lead to lots of silent correctness issues.

Reply:
Perhaps allow for the `identity = infty` behaviour with a flag? Definitely not in this PR, but opening an issue about for a future PR could be interesting.